### PR TITLE
Mcp skills

### PR DESCRIPTION
**Pull request by v-aidaba**

### CODE REVIEW

File: src/mcp/tools/skills.ts [hunk 1]
```diff
@@ -0,0 +1,178 @@
+/*
```
Comment (Demonkratiy) [inline src/mcp/tools/skills.ts:1]: **Missing tests for `skills.ts`**

All existing MCP tools have unit tests in `spec/unit/McpToolsSpec.js`. This new module has no test coverage.

Suggested approach:
- Unit tests with mocked `fetch` to cover: manifest parsing, skill lookup, error handling for unknown skills, network failure graceful degradation, and caching behavior.
- Optionally a single integration smoke test (skippable without network) that validates the real remote `skills.json` is parseable.
- Export a cache reset helper (e.g. `_resetCache()`) to isolate tests from each other.

File: src/mcp/tools/skills.ts [hunk 2]
```diff
@@ -0,0 +1,178 @@
...
+    path: string;
+    type: "file" | "dir";
+    download_url: string | null;
+}
+
+const REPO_OWNER = "Demonkratiy";
+const REPO_NAME = "powerbi-visuals-skills";
+const BRANCH = "main";
+const RAW_BASE = `https://raw.githubusercontent.com/${REPO_OWNER}/${REPO_NAME}/${BRANCH}`;
+const API_BASE = `https://api.github.com/repos/${REPO_OWNER}/${REPO_NAME}/contents`;
```
Comment (Demonkratiy) [inline src/mcp/tools/skills.ts]: **Rate limiting risk: consider removing `api.github.com` dependency**

`api.github.com` has a strict 60 req/hr limit for unauthenticated requests. `fetchDirListing` + `collectFilesRecursively` can consume multiple requests per single `implement_feature` call.

Suggested solution: in the root `skills.json` manifest, extend each skill descriptor with a `files` array listing all file paths relative to the skill directory. Example:
`json
{ "id": "bookmarks", "path": "skills/bookmarks", "files": ["SKILL.md", "references/capabilities.json"] }
`
Then fetch all content exclusively via `raw.githubusercontent.com` using known paths — no directory listing needed. This eliminates the `api.github.com` dependency entirely.

Don't forget to update the related documentation in the `powerbi-visuals-skills` repository (README, CONTRIBUTING, schema description) if adding the `files` field.

If keeping `api.github.com` short-term: at minimum, catch 403/429 responses and show a user-friendly message with reset time from `X-RateLimit-Reset` header.

File: src/mcp/tools/skills.ts [hunk 3]
```diff
@@ -0,0 +1,178 @@
...
+
+let cachedManifest: SkillsManifest | null = null;
+
+async function fetchRawFile(repoPath: string): Promise<string> {
```
Comment (Demonkratiy) [inline src/mcp/tools/skills.ts:47]: **Manifest cache lives forever (no TTL)**

The MCP server process lives as long as the IDE session (STDIO transport, killed only on disconnect/SIGINT). This can be hours or days. `cachedManifest` is never invalidated, so newly published skills won't appear until the user restarts the IDE.

Suggestion: add a TTL (e.g. 30-60 min) — store a timestamp alongside the cached value and re-fetch if expired.

File: src/mcp/tools/skills.ts [hunk 4]
```diff
@@ -0,0 +1,178 @@
...
+            if (entry.type === "file" && entry.download_url) {
+                const content = await fetchRawFile(entry.path);
+                files.push({ filename: entry.name, content });
+            } else if (entry.type === "dir") {
+                const subFiles = await collectFilesRecursively(entry.path);
+                files.push(...subFiles);
+            }
+        }
+    } catch {
+        // If listing fails (e.g., rate limit), return what we have
```
Comment (Demonkratiy) [inline src/mcp/tools/skills.ts]: **Silent error swallowing — user gets incomplete results without notice**

If fetching additional skill files fails, the error is silently discarded. The user receives a partial response with no indication that something is missing.

Regardless of the fetch mechanism used, when individual files fail to load: include a warning in the response (e.g. `"warnings": ["Failed to fetch references/capabilities.json"]`) so the AI agent and user are aware the response is incomplete.

File: src/mcp/McpServer.ts [hunk 5]
```diff
@@ -20,17 +20,51 @@ import { checkVulnerabilities } from "./tools/vulnerabilities.js";
...
-            name: "pbiviz-mcp-server",
-            version: "1.0.0",
-        });
+        this.server = new MCPServerSDK(
+            {
+                name: "pbiviz-mcp-server",
+                version: "1.0.0",
+            },
+            {
+                instructions:
```
Comment (Demonkratiy) [inline src/mcp/McpServer.ts:37]: **Overloaded server instructions — consider a concise semantic formulation**

LLMs should understand intent semantically — they don't do keyword matching. Listing every possible phrase and spelling variation ("bookmrks", "toltips") likely doesn't improve tool selection — it dilutes the signal and inflates token usage on every exchange.

Consider replacing the 30+ line instructions with a short role-based formulation, e.g.:
```
"MCP server for Power BI custom visual development. Provides tools for: implementing visual features, checking best practices and security, certification readiness, project info, and available APIs. Use these tools whenever the user's request relates to Power BI visual development."
```
The tool selection guide (which tool for which intent) is better conveyed by each tool's own description.

File: src/mcp/McpServer.ts [hunk 6]
```diff
@@ -39,7 +73,13 @@ export class McpServer {
         // Tool 1: Get Best Practices
         this.server.tool(
             "get_best_practices",
-            "Returns best practice guidelines for Power BI custom visual development. Covers: API version management, performance optimization (update loop, lazy loading, data processing), security (eval, innerHTML, XSS, sanitization, external call/calls, network request/requests), accessibility (keyboard navigation, high contrast, screen reader, ARIA label/labels), project structure (module/modules, error handling), formatting pane (format model, formatting model), testing (unit test/tests, E2E test/tests, edge case/cases), and documentation (README, changelog, comment/comments).",
+            "MUST CALL when user asks about Power BI visual quality, standards, or how to build correctly. " +
+            "Get best practices, coding guidelines, and recommendations for building Power BI custom visuals. " +
```
Comment (Demonkratiy) [inline src/mcp/McpServer.ts]: **Tool descriptions: prefer semantic clarity over keyword lists**

"MUST CALL", "Trigger phrases: ..." and keyword enumerations likely don't help LLMs — they should understand natural language intent. A clear, concise description of *what the tool does and when to use it* might work better and save tokens.

Example — instead of the current 6-line description, something like:
```
"Get best practices and coding guidelines for Power BI custom visual development. Covers performance, security, accessibility, project structure, and testing."
```
Apply the same principle to all tool descriptions.

File: src/mcp/McpServer.ts [hunk 7]
```diff
@@ -102,6 +169,56 @@ export class McpServer {
...
+            "Call this tool when the user wants to: add, implement, enable, integrate, support, use, or ask about ANY feature or capability. " +
+            "Trigger words: bookmarks, bookmark, tooltips, tooltip, context menu, drill-down, drill down, drilldown, " +
+            "dialog, dialog box, modal, warning icon, warning, landing page, formatting, format pane, " +
+            "selection, highlight, high contrast, localization, local storage, storage, file download, download, " +
+            "rendering events, events, keyboard navigation, keyboard, analytics, analytics pane, " +
+            "conditional formatting, sync slicer, slicer, launch URL, URL, advanced edit, edit mode, " +
+            "total, subtotal, sub-total, allow interactions, interactions, fetch more data, color palette. " +
+            "Also call when user says: 'support X', 'implement X', 'enable X', 'how to add X', 'I need X', 'integrate X'. " +
+            "Even with typos like 'bookmrks', 'toltips', 'contex menu' — STILL call this tool. " +
+            "Returns feature IDs — then call get_feature_documentation with the ID for full instructions.",
```
Comment (Demonkratiy) [inline src/mcp/McpServer.ts]: Bug: references non-existent tool `get_feature_documentation`. Should be `implement_feature`.

File: src/mcp/tools/skills.ts [hunk 8]
```diff
@@ -0,0 +1,178 @@
...
+
+export async function getSkillInstructions(skillName: string): Promise<string> {
+    const manifest = await getManifest();
+    const skill = findSkill(manifest, skillName);
+
+    if (!skill) {
+        const available = manifest.skills.map(s => s.id);
+        return JSON.stringify({
+            error: `Feature "${skillName}" not found.`,
+            message: `Please pick a valid feature ID from the list below and call get_feature_documentation again.`,
```
Comment (Demonkratiy) [inline src/mcp/tools/skills.ts]: Bug: references non-existent tool `get_feature_documentation`. Should be `implement_feature`.

File: MCP.md [hunk 9]
```diff
@@ -57,11 +57,13 @@ The server runs on STDIO, so you won't see output, but it will respond to MCP re
...
-| `check_vulnerabilities` | Scans project for security vulnerabilities in dependencies and code | ✅ | ✅ |
-| `prepare_certification` | Audits visual for Power BI certification readiness | ✅ | ✅ |
-| `list_visual_info` | Returns info about current visual (name, GUID, API version, capabilities) | ✅ | ✅ |
-| `get_available_apis` | Lists available Power BI Visual APIs with examples | ✅ | ✅ |
+| `get_best_practices` | Get best practices and guidelines for Power BI visual development | ✅ | ✅ |
+| `check_vulnerabilities` | Scan project for security vulnerabilities and dangerous code patterns | ✅ | ✅ |
+| `prepare_certification` | Check if the visual is ready for certification and marketplace submission | ✅ | ✅ |
+| `list_visual_info` | Get info about current visual (name, GUID, API version, capabilities) | ✅ | ✅ |
+| `get_available_apis` | List available Power BI Visual APIs with examples | ✅ | ✅ |
+| `add_feature` | List all features that can be added to a Power BI visual | ✅ | ✅ |
```
Comment (Demonkratiy) [inline MCP.md]: `add_feature` and `implement_feature` fetch data from a remote GitHub repository. The "Local" column should be ❌ for these tools (at minimum `add_feature` requires a remote request at least once per session to fetch the manifest). They don't send any user data externally, but they do require network access to function.

File: src/mcp/tools/skills.ts (hunk 1 above)
Comment (v-aidaba) [reply src/mcp/tools/skills.ts:1]: Done. Added a Skills describe block in McpToolsSpec.js with 7 unit tests.

File: src/mcp/tools/skills.ts (hunk 2 above)
Comment (v-aidaba) [reply src/mcp/tools/skills.ts]: Done. skills.ts no longer references api.github.com ,there's no API_BASE, no fetchDirListing, and no contents-API calls.

File: src/mcp/tools/skills.ts (hunk 3 above)
Comment (v-aidaba) [reply src/mcp/tools/skills.ts:47]: Done

File: src/mcp/tools/skills.ts (hunk 4 above)
Comment (v-aidaba) [reply src/mcp/tools/skills.ts]: Done. getAdditionalSkillFiles now collects per-file failures into a warnings

File: src/mcp/McpServer.ts (hunk 5 above)
Comment (v-aidaba) [reply src/mcp/McpServer.ts:37]: Changed

File: src/mcp/McpServer.ts (hunk 6 above)
Comment (v-aidaba) [reply src/mcp/McpServer.ts]: Fixed

File: src/mcp/McpServer.ts (hunk 7 above)
Comment (v-aidaba) [reply src/mcp/McpServer.ts]: Fixed

File: src/mcp/tools/skills.ts (hunk 8 above)
Comment (v-aidaba) [reply src/mcp/tools/skills.ts]: Fixed

File: MCP.md (hunk 9 above)
Comment (v-aidaba) [reply MCP.md]: Fixed

File: src/mcp/tools/skills.ts (hunk 1 above)
Comment (Demonkratiy) [reply src/mcp/tools/skills.ts:1]: Resolved

File: src/mcp/tools/skills.ts (hunk 3 above)
Comment (Demonkratiy) [reply src/mcp/tools/skills.ts:47]: Resolved

File: src/mcp/tools/skills.ts (hunk 4 above)
Comment (Demonkratiy) [reply src/mcp/tools/skills.ts]: Resolved

File: src/mcp/McpServer.ts (hunk 5 above)
Comment (Demonkratiy) [reply src/mcp/McpServer.ts:37]: Resolved

File: src/mcp/McpServer.ts (hunk 6 above)
Comment (Demonkratiy) [reply src/mcp/McpServer.ts]: Resolved

File: src/mcp/McpServer.ts (hunk 7 above)
Comment (Demonkratiy) [reply src/mcp/McpServer.ts]: Resolved

File: src/mcp/tools/skills.ts (hunk 8 above)
Comment (Demonkratiy) [reply src/mcp/tools/skills.ts]: Resolved

File: MCP.md (hunk 9 above)
Comment (Demonkratiy) [reply MCP.md]: Resolved

File: src/mcp/tools/skills.ts (hunk 2 above)
Comment (Demonkratiy) [reply src/mcp/tools/skills.ts]: **Reopening - fix is incomplete: `skills.json` in `powerbi-visuals-skills` project is not updated**

Verified against the live remote repo: the `files` array has not been added to any skill in `skills.json`. As a result, `implement_feature` currently returns only `SKILL.md` plus a warning, and the actual implementation guide (`references/README.md` with code samples, checklists, common mistakes) is **never loaded**.

Tested locally:
```
implement_feature('bookmarks-support')
{ "instructions": "...See [references/README.md](references/README.md)...",
  "warnings": ["Skill \"bookmarks-support\" manifest entry is missing a \"files\" array; no reference files will be loaded."] }
```

The `references/README.md` file exists in the repo and contains the bulk of the implementation content.

Before merging: update `skills.json` in `powerbi-visuals-skills` project to include the `files` array for every skill (listing `SKILL.md` + all `references/*` files).

File: spec/unit/McpToolsSpec.js [hunk 10]
```diff
@@ -412,4 +413,150 @@
...
+        it("-> should never call api.github.com (raw-only fetching)", async () => {
+            mockFetch({
+                'skills.json': mockManifest,
+                'skills/bookmarks/SKILL.md': '# Bookmarks',
+                'references/capabilities.json': '{}'
+            });
+
+            await getSkillInstructions('bookmarks');
+
+            const apiCalls = globalThis.fetch.calls.allArgs().filter(args => args[0].includes('api.github.com'));
```
[the remark this answers is not in the document]
Comment (v-aidaba) [reply spec/unit/McpToolsSpec.js]: Fixed

File: src/mcp/tools/skills.ts (hunk 2 above)
Comment (Demonkratiy) [reply src/mcp/tools/skills.ts]: Verified - `skills.json` now contains `files` for every skill, and `implement_feature` correctly returns the full `references/README.md` content. Resolved.

**Review by Demonkratiy — APPROVED** [review]
